File: NOVEMBER-2025/10-11-2025/Stock Buy and Sell with Cooldown.js
/**
 * @param {number[]} arr
 * @returns {number}
 */

class Solution {
    maxProfit(arr) {
        const n = arr.length;
        if (n === 0) return 0;
        let hold = -arr[0];
        let sold = 0;
        let rest = 0;
        for (let i = 1; i < n; i++) {
            const prevHold = hold;
            const prevSold = sold;
            const prevRest = rest;
            hold = Math.max(prevHold, prevRest - arr[i]);
            sold = prevHold + arr[i];
            rest = Math.max(prevRest, prevSold);
        }
        return Math.max(sold, rest);
    }
}

File: NOVEMBER-2025/10-11-2025/Stock Buy and Sell with Cooldown.cs
class Solution {
    public int maxProfit(int[] arr) {
        int n = arr.Length;
        if (n == 0) return 0;
        int hold = -arr[0];
        int sold = 0;
        int rest = 0;
        for (int i = 1; i < n; i++) {
            int prevHold = hold;
            int prevSold = sold;
            int prevRest = rest;
            hold = Math.Max(prevHold, prevRest - arr[i]);
            sold = prevHold + arr[i];
            rest = Math.Max(prevRest, prevSold);
        }
        return Math.Max(sold, rest);
    }
}